### DISCUSSION TITLE
How to view `defer` loading state, similar to `loader` with `useNavigate`

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [remix-run/remix#7563] [Q&A]
**Background Context**

For example I know with loader you can use useNavigate to see if some data is in progress

```tsx
const SomeComponent = () => {
  const navigation = useNavigation();
  const isLoading = navigation.state !== 'idle'
  return (<p>Are we loading? :{isLoading}</>)
}
```

However, is there a similar way to hook into the progress of data that is coming from `defer`?
Use case: implement a spinner outside of where the stream'd data is rendered

```tsx
const SomeComponentWhereDeferDataWillRender = () => {
  return (
     <Suspense fallback={<div>Loading...</div>}>
      <Await resolve={somePromise}> {(resolvedValue) => <p>{resolvedValue}</p>}</Await>
     </Suspense>;
  )
}

const PageComponent = () => {
   const isLoading = /*somehow hook into the defer?*/
  
   return (
    <TopNav isLoading={isLoading} />
    <SomeComponentWhereDeferDataWillRender />
  );
}
 
```

## When I hook into `useNavigate` it doesnt seem to update when I make a page transition:
- Home page ---> Page with `defer`
- useNavigate doesnt update, it remains `idle`
[Stackblitz Code ](https://stackblitz.com/edit/node-2bddzb?file=package.json,app%2Froutes%2Fstreaming-ssr-route3-mix-critical-and-non-critical-data.tsx)

https://github.com/remix-run/remix/assets/6743796/9f10b17c-9baf-4e45-91a1-e856629d8d2e

## References
- [Remix Discord Thread](https://discord.com/channels/770287896669978684/1157852129516597323)
- [useNavigate](https://remix.run/docs/en/main/hooks/use-navigate)
- [defer](https://remix.run/docs/en/main/utils/defer)
- [stackblitz example](https://stackblitz.com/edit/node-2bddzb?file=package.json,app%2Froutes%2Fstreaming-ssr-route3-mix-critical-and-non-critical-data.tsx)

**Comment by isc30** [comment]
same question here, havent found a solution :(

**Comment by melgrove** [comment]
The main issue here is that the deferred value (promise) does not trigger a state update when it resolves. I had the same problem and I've achieved this by using useEffect and awaiting the promise directly. In the original example it would be like this:

```jsx
const SomeComponentWhereDeferDataWillRender = ({ setIsLoading }) => {
  useEffect(() => {
    (async () => {
      await somePromise;
      setIsLoading(false);
    })();
  }, [setIsLoading, somePromise]);

  return (
    <Suspense fallback={<div>Loading...</div>}>
    <Await resolve={somePromise}> {(resolvedValue) => <p>{resolvedValue}</p>}</Await>
    </Suspense>
  );
}

const PageComponent = () => {
  const [isLoading, setIsLoading] = useState(true);
  
  return (
    <TopNav isLoading={isLoading} />
    <SomeComponentWhereDeferDataWillRender setIsLoading={setIsLoading} />
  );
}
```
